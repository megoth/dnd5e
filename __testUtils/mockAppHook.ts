import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
