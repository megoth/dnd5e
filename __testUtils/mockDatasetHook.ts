import { mockSolidDatasetFrom, SolidDataset } from "@inrupt/solid-client";
import { SWRResponse } from "swr";

export default function mockDatasetHook(
  mock,
  presets: Partial<SWRResponse<SolidDataset, any>> = {}
) {
  return mock.mockReturnValue({
    data: mockSolidDatasetFrom("http://example.com"),
    error: null,
    ...presets,
  });
}
