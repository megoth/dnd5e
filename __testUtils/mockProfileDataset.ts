import {
  addStringNoLocale,
  addUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { vcard, foaf, rdf } from "rdf-namespaces";
import { chain } from "../src/utils";

export const aliceWebIdDatasetUrl = "http://example.com/alice";
export const aliceWebIdUrl = `${aliceWebIdDatasetUrl}#me`;
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://example.com/alice.jpg";

export function mockProfileDataset(webId = aliceWebIdUrl) {
  const profile = chain(
    mockThingFrom(webId),
    (t) => addStringNoLocale(t, vcard.fn, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, foaf.openid, aliceWebIdUrl)
  );
  return chain(mockSolidDatasetFrom(aliceWebIdDatasetUrl), (d) =>
    setThing(d, profile)
  );
}
