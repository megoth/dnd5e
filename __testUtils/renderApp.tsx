import React from "react";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getFluentBundles } from "../src/models/language";
import { AppModel } from "../src/models/app";

export default function renderApp(
  app: AppModel,
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult {
  function rerender() {
    return render(
      <LocalizationProvider l10n={new ReactLocalization(getFluentBundles(app))}>
        {ui}
      </LocalizationProvider>,
      options
    );
  }
  const renderResult = rerender();
  return { ...renderResult, rerender };
}
