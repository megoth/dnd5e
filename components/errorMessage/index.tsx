import React from "react";
import { Localized } from "@fluent/react";
import NestedError from "nested-error-stacks";
import { getErrorId } from "../../src/models/error";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";
export const TESTID_ERROR_STACK = "error-stack";

export default function ErrorMessage({ error }: Props) {
  const { message: errorUrl } = error;
  const errorId = getErrorId(errorUrl);
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        <Localized id={errorId}>{errorUrl}</Localized>
      </h1>
      <pre data-testid={TESTID_ERROR_STACK}>
        <h2>{error.message}</h2>
        <pre>{error.stack}</pre>
      </pre>
    </div>
  );
}
