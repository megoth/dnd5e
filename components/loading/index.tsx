import React from "react";

export const TESTID_LOADING = "loading";

export default function Loading() {
  // Do not try to translate anything here - it's used when loading translations as well
  return <div data-testid={TESTID_LOADING}>Loading...</div>;
}
