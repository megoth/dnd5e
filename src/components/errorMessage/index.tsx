import React from "react";
import NestedError from "nested-error-stacks";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  // const { message: errorURL } = error;
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        {/*{translationURL ? (*/}
        {/*  <Localized id={getTranslationId(translationURL)}>*/}
        {/*    {errorURL}*/}
        {/*  </Localized>*/}
        {/*) : (*/}
        {error.message}
        {/*)}*/}
      </h1>
    </div>
  );
}
