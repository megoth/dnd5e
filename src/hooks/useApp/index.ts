import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
