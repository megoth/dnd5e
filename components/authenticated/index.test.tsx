import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as solidUIReactFns from "@inrupt/solid-ui-react";
import Authenticated, { TESTID_AUTHENTICATED_LOGOUT_BUTTON } from "./index";
import { mockProfileDataset } from "../../__testUtils/mockProfileDataset";
import mockDatasetHook from "../../__testUtils/mockDatasetHook";
import useDataset from "../../src/hooks/useDataset";
import { TESTID_LOADING } from "../loading";
import { TESTID_ERROR } from "../errorMessage";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import {
  authenticatedWebId,
  mockAuthenticatedSession,
} from "../../__testUtils/mockSession";
import renderApp from "../../__testUtils/renderApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

jest.mock("../../src/hooks/useDataset");
const mockedDatasetHook = useDataset as jest.Mock;

describe("Authenticated", () => {
  const data = mockProfileDataset(authenticatedWebId);
  const authenticatedSession = mockAuthenticatedSession();

  let app;

  beforeEach(() => {
    app = mockAppHook(mockedAppHook);
  });
  beforeEach(() => {
    jest
      .spyOn(solidUIReactFns, "useSession")
      .mockReturnValue(authenticatedSession);
  });

  it("renders info about the user and a log out button", () => {
    mockDatasetHook(mockedDatasetHook, { data });
    const { asFragment, queryByTestId } = renderApp(app, <Authenticated />);
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId(TESTID_LOADING)).toBeNull();
    expect(queryByTestId(TESTID_ERROR)).toBeNull();
  });

  it("renders loading while fetching profile", () => {
    mockDatasetHook(mockedDatasetHook, { data: null });
    const { asFragment, getByTestId } = renderApp(app, <Authenticated />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_LOADING)).toBeDefined();
  });

  it("renders error if fetching profile fails", () => {
    mockDatasetHook(mockedDatasetHook, { error: new Error() });
    const { asFragment, getByTestId } = renderApp(app, <Authenticated />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_ERROR)).toBeDefined();
  });

  it("renders a logout button", () => {
    mockDatasetHook(mockedDatasetHook, { data });
    const { logout } = authenticatedSession;
    const { getByTestId } = renderApp(app, <Authenticated />);
    userEvent.click(getByTestId(TESTID_AUTHENTICATED_LOGOUT_BUTTON));
    expect(logout).toHaveBeenCalledWith();
  });
});
