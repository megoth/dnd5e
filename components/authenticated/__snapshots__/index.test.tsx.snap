// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticated renders error if fetching profile fails 1`] = `
<DocumentFragment>
  <div
    data-testid="error"
  >
    <h1
      data-testid="error-title"
    />
    <pre
      data-testid="error-stack"
    >
      <h2 />
      <pre>
        Error: 
    at Object.&lt;anonymous&gt; (/Users/arnehassel/Projects/dnd5e-app/components/authenticated/index.test.tsx:31:49)
    at Object.asyncJestTest (/Users/arnehassel/Projects/dnd5e-app/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)
    at /Users/arnehassel/Projects/dnd5e-app/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise (&lt;anonymous&gt;)
    at mapper (/Users/arnehassel/Projects/dnd5e-app/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /Users/arnehassel/Projects/dnd5e-app/node_modules/jest-jasmine2/build/queueRunner.js:75:41
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
      </pre>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`Authenticated renders info about the user and a log out button 1`] = `
<DocumentFragment>
  <p />
  <div
    role="button"
    tabindex="0"
  >
    <button
      type="button"
    />
  </div>
</DocumentFragment>
`;

exports[`Authenticated renders loading while fetching profile 1`] = `
<DocumentFragment>
  <div
    data-testid="loading"
  >
    Loading...
  </div>
</DocumentFragment>
`;
